File: web/assets/themes/default.go
package themes

import (
	_ "embed"
	"encoding/json"
	"fmt"
	"strconv"

	v1 "github.com/turnforge/weewar/gen/go/weewar/v1/models"
)

//go:embed default/mapping.json
var defaultMappingJSON []byte

// ThemeMappingEntry represents a single unit or terrain entry in mapping.json
type ThemeMappingEntry struct {
	Old   string `json:"old"`
	Name  string `json:"name"`
	Image string `json:"image"`
}

// PlayerColorJSON represents a player color entry in mapping.json
type PlayerColorJSON struct {
	Primary   string `json:"primary"`
	Secondary string `json:"secondary"`
	Name      string `json:"name"`
}

// ThemeInfo from mapping.json
type ThemeInfoJSON struct {
	Name                string `json:"name"`
	Version             string `json:"version"`
	BasePath            string `json:"base_path"`
	AssetType           string `json:"asset_type"`
	NeedsPostProcessing bool   `json:"needs_post_processing"`
}

// ThemeMapping represents the full mapping.json structure
type ThemeMappingJSON struct {
	ThemeInfo    ThemeInfoJSON                `json:"themeInfo"`
	Units        map[string]ThemeMappingEntry `json:"units"`
	Terrains     map[string]ThemeMappingEntry `json:"terrains"`
	PlayerColors map[string]PlayerColorJSON   `json:"playerColors"`
}

// DefaultTheme implements the Theme interface for PNG-based default assets
// Loads data from embedded mapping.json
type DefaultTheme struct {
	themeInfo    ThemeInfoJSON
	units        map[int32]ThemeMappingEntry
	terrains     map[int32]ThemeMappingEntry
	playerColors map[int32]*v1.PlayerColor
	cityTerrains map[int32]bool // Terrains that use player colors (from RulesEngine)
}

// NewDefaultTheme creates a new default theme instance by parsing embedded mapping.json
// cityTerrains is a map of terrain IDs that use player colors (from RulesEngine.TerrainTypes)
func NewDefaultTheme(cityTerrains map[int32]bool) *DefaultTheme {
	var mapping ThemeMappingJSON
	if err := json.Unmarshal(defaultMappingJSON, &mapping); err != nil {
		panic(fmt.Sprintf("failed to parse embedded default theme mapping: %v", err))
	}

	// Convert string keys to int32 for units
	units := make(map[int32]ThemeMappingEntry)
	for key, entry := range mapping.Units {
		id, err := strconv.ParseInt(key, 10, 32)
		if err == nil {
			units[int32(id)] = entry
		}
	}

	// Convert string keys to int32 for terrains
	terrains := make(map[int32]ThemeMappingEntry)
	for key, entry := range mapping.Terrains {
		id, err := strconv.ParseInt(key, 10, 32)
		if err == nil {
			terrains[int32(id)] = entry
		}
	}

	// Convert string keys to int32 for player colors
	playerColors := make(map[int32]*v1.PlayerColor)
	for key, entry := range mapping.PlayerColors {
		id, err := strconv.ParseInt(key, 10, 32)
		if err == nil {
			playerColors[int32(id)] = &v1.PlayerColor{
				Primary:   entry.Primary,
				Secondary: entry.Secondary,
			}
		}
	}

	return &DefaultTheme{
		themeInfo:    mapping.ThemeInfo,
		units:        units,
		terrains:     terrains,
		playerColors: playerColors,
		cityTerrains: cityTerrains,
	}
}

func (d *DefaultTheme) GetUnitName(unitId int32) string {
	if entry, ok := d.units[unitId]; ok {
		return entry.Name
	}
	return ""
}

func (d *DefaultTheme) GetTerrainName(terrainId int32) string {
	if entry, ok := d.terrains[terrainId]; ok {
		return entry.Name
	}
	return ""
}

func (d *DefaultTheme) GetUnitDescription(unitId int32) string {
	// Default theme has no custom descriptions
	return ""
}

func (d *DefaultTheme) GetTerrainDescription(terrainId int32) string {
	// Default theme has no custom descriptions
	return ""
}

// GetUnitPath returns the directory path for a unit's assets
// For PNG themes, this is the directory containing player-colored variants
func (d *DefaultTheme) GetUnitPath(unitId int32) string {
	if entry, ok := d.units[unitId]; ok {
		return fmt.Sprintf("%s/%s", d.themeInfo.BasePath, entry.Image)
	}
	return ""
}

// GetTilePath returns the directory path for a terrain's assets
func (d *DefaultTheme) GetTilePath(terrainId int32) string {
	if entry, ok := d.terrains[terrainId]; ok {
		return fmt.Sprintf("%s/%s", d.themeInfo.BasePath, entry.Image)
	}
	return ""
}

// GetUnitAssetPath returns the full path to a specific unit+player PNG file
func (d *DefaultTheme) GetUnitAssetPath(unitId, playerId int32) string {
	if entry, ok := d.units[unitId]; ok {
		return fmt.Sprintf("%s/%s/%d.png", d.themeInfo.BasePath, entry.Image, playerId)
	}
	return ""
}

// GetTileAssetPath returns the full path to a specific terrain+player PNG file
func (d *DefaultTheme) GetTileAssetPath(terrainId, playerId int32) string {
	if entry, ok := d.terrains[terrainId]; ok {
		// Only city terrains use player colors; all others use player 0 (neutral)
		effectivePlayer := int32(0)
		if d.cityTerrains[terrainId] {
			effectivePlayer = playerId
		}
		return fmt.Sprintf("%s/%s/%d.png", d.themeInfo.BasePath, entry.Image, effectivePlayer)
	}
	return ""
}

func (d *DefaultTheme) GetThemeInfo() *v1.ThemeInfo {
	return &v1.ThemeInfo{
		Name:                d.themeInfo.Name,
		Version:             d.themeInfo.Version,
		BasePath:            d.themeInfo.BasePath,
		AssetType:           d.themeInfo.AssetType,
		NeedsPostProcessing: d.themeInfo.NeedsPostProcessing,
	}
}

func (d *DefaultTheme) GetAvailableUnits() []int32 {
	units := make([]int32, 0, len(d.units))
	for id := range d.units {
		units = append(units, id)
	}
	return units
}

func (d *DefaultTheme) GetAvailableTerrains() []int32 {
	terrains := make([]int32, 0, len(d.terrains))
	for id := range d.terrains {
		terrains = append(terrains, id)
	}
	return terrains
}

func (d *DefaultTheme) HasUnit(unitId int32) bool {
	_, ok := d.units[unitId]
	return ok
}

func (d *DefaultTheme) HasTerrain(terrainId int32) bool {
	_, ok := d.terrains[terrainId]
	return ok
}

func (d *DefaultTheme) GetEffectivePlayer(terrainId, playerId int32) int32 {
	if d.cityTerrains[terrainId] {
		return playerId
	}
	return 0
}

func (d *DefaultTheme) GetPlayerColor(playerId int32) *v1.PlayerColor {
	if color, ok := d.playerColors[playerId]; ok {
		return color
	}
	// Fallback to neutral if not found
	if color, ok := d.playerColors[0]; ok {
		return color
	}
	return nil
}

// GetAssetPathForTemplate is a helper for templates to get either unit or tile paths
func (d *DefaultTheme) GetAssetPathForTemplate(assetType string, assetId, playerId int32) string {
	switch assetType {
	case "unit":
		return d.GetUnitAssetPath(assetId, playerId)
	case "tile", "terrain":
		return d.GetTileAssetPath(assetId, playerId)
	default:
		return ""
	}
}

File: web/assets/themes/default.ts
/**
 * Default Theme Provider
 * Extends BaseTheme with PNG-specific asset loading
 * Uses pre-colored PNG assets that don't need post-processing
 */

import { BaseTheme, ThemeInfo, ThemeMapping, PlayerColor } from './BaseTheme';
import mappingData from './default/mapping.json';

// Type for the extended mapping with playerColors
interface DefaultMappingData extends ThemeMapping {
    themeInfo: {
        name: string;
        version: string;
        base_path: string;
        asset_type: string;
        needs_post_processing: boolean;
    };
    playerColors: { [key: string]: PlayerColor };
}

const mapping = mappingData as DefaultMappingData;

/**
 * Default Theme Implementation
 * Uses PNG assets with pre-colored player variants
 */
export default class DefaultTheme extends BaseTheme {
    protected basePath = '/static/assets/themes/default';
    protected themeName = 'Default (PNG)';
    protected themeVersion = '1.0.0';

    constructor() {
        super(mapping);
    }

    /**
     * Override getThemeInfo for PNG-specific settings
     */
    getThemeInfo(): ThemeInfo {
        return {
            name: this.themeName,
            version: this.themeVersion,
            basePath: this.basePath,
            supportsTinting: false,
            needsPostProcessing: false,
            assetType: 'png',
            playerColors: this.playerColors as any,
        };
    }

    /**
     * Override getUnitPath - PNG assets don't use template paths
     */
    getUnitPath(unitId: number): string | undefined {
        // PNG assets use getUnitAssetPath instead
        return undefined;
    }

    /**
     * Override getTilePath - PNG assets don't use template paths
     */
    getTilePath(terrainId: number): string | undefined {
        // PNG assets use getTileAssetPath instead
        return undefined;
    }

    /**
     * Returns the direct path to a pre-colored unit asset
     */
    getUnitAssetPath(unitId: number, playerId: number): string | undefined {
        const unit = this.unitMapping[unitId.toString()];
        if (!unit) {
            return undefined;
        }
        // PNG assets are stored as: {basePath}/{image}/{playerId}.png
        // where image is a directory like "Units/15"
        return `${this.basePath}/${unit.image}/${playerId}.png`;
    }

    /**
     * Returns the direct path to a pre-colored terrain asset
     */
    getTileAssetPath(terrainId: number, playerId: number): string | undefined {
        const terrain = this.terrainMapping[terrainId.toString()];
        if (!terrain) {
            return undefined;
        }

        // Only city terrains use player colors; all others use player 0 (neutral)
        const effectivePlayer = this.isCityTile(terrainId) ? playerId : 0;
        // PNG assets are stored as: {basePath}/{image}/{playerId}.png
        // where image is a directory like "Tiles/5"
        return `${this.basePath}/${terrain.image}/${effectivePlayer}.png`;
    }

    /**
     * Override loadUnit - PNG theme uses pre-colored assets
     */
    async loadUnit(unitId: number, playerId: number): Promise<string> {
        throw new Error('Default theme uses pre-colored PNG assets. Use getUnitAssetPath() instead.');
    }

    /**
     * Override loadTile - PNG theme uses pre-colored assets
     */
    async loadTile(terrainId: number, playerId?: number): Promise<string> {
        throw new Error('Default theme uses pre-colored PNG assets. Use getTileAssetPath() instead.');
    }

    /**
     * Override setUnitImage for PNG assets
     */
    async setUnitImage(unitId: number, playerId: number, targetElement: HTMLElement): Promise<void> {
        const path = this.getUnitAssetPath(unitId, playerId);
        if (!path) {
            targetElement.innerHTML = '⚔️';
            return;
        }

        targetElement.innerHTML = '';
        const img = document.createElement('img');
        img.src = path;
        img.alt = this.getUnitName(unitId) || `Unit ${unitId}`;
        img.className = 'w-full h-full object-contain';
        img.style.imageRendering = 'pixelated';

        img.onerror = () => {
            targetElement.innerHTML = '⚔️';
        };

        targetElement.appendChild(img);
    }

    /**
     * Override setTileImage for PNG assets
     */
    async setTileImage(tileId: number, playerId: number, targetElement: HTMLElement): Promise<void> {
        const path = this.getTileAssetPath(tileId, playerId);
        if (!path) {
            targetElement.innerHTML = '🏞️';
            return;
        }

        targetElement.innerHTML = '';
        const img = document.createElement('img');
        img.src = path;
        img.alt = this.getTerrainName(tileId) || `Terrain ${tileId}`;
        img.className = 'w-full h-full object-contain';
        img.style.imageRendering = 'pixelated';

        img.onerror = () => {
            targetElement.innerHTML = '🏞️';
        };

        targetElement.appendChild(img);
    }
}
